Loggers

torchiaf opened the pull request:
- Add logger utilities
- Enable logs - Rest client

Output

![image](https://github.com/torchiaf/kubectl-rancherx/assets/26394656/eb50a0a9-a128-4c50-b2f4-f6b9cb73c695)

File

![image](https://github.com/torchiaf/kubectl-rancherx/assets/26394656/10b7251c-dfb4-492c-9e7d-1a1b6c179de0)

torchiaf committed:
Add logger; Add debug flag to rootCmd

Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

torchiaf committed:
Enable logging - rest client

Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

torchiaf commented on pkg/manager/rest.go:
```diff
@@ -21,9 +22,12 @@ func List[T runtime.Object](ctx context.Context, client *rest.RESTClient, resour
 		Into(obj)
 
 	if err != nil {
+		log.Tracef("manager.List -  resource [%s] - namespace [%s] - error [%s]", resource, namespace, err)
```
What would be the best practice to make this trace using a variable number of parameters in Go? @enrichman

torchiaf replied:
See `Logger.with()` https://go.dev/blog/slog

torchiaf commented on pkg/cli/logger.go:
```diff
@@ -0,0 +1,44 @@
+
+	// Setup the log output
+	currentTime := time.Now()
+	logFile := fmt.Sprintf("rancherx-%d-%d-%d.log", currentTime.Year(), currentTime.Month(), currentTime.Day())
```
Create a new flag to explicity create the log file (specify a name).

torchiaf commented on pkg/cli/logger.go:
```diff
@@ -0,0 +1,44 @@
+	"os"
+	"time"
+
+	log "github.com/sirupsen/logrus"
```
To use in combination with official `slog` package

torchiaf commented on pkg/manager/rest.go:
```diff
@@ -21,9 +22,12 @@ func List[T runtime.Object](ctx context.Context, client *rest.RESTClient, resour
 		return err
 	}
 
+	log.Tracef("manager.List - resource [%s] - namespace [%s] - obj [%+v]", resource, namespace, obj)
```
Try to re-define `string()` method for specialized types.

torchiaf committed:
Decouple log level from print to log file; add [-p fileName]flag

Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

torchiaf committed:
Replace logrus with slog

Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

torchiaf committed:
Add custom levels: Trace, Fatal

Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

torchiaf committed:
Format rest trace errors

Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

enrichman commented on pkg/cli/cli.go:
```diff
@@ -16,6 +18,12 @@ func NewRootCmd() (*cobra.Command, error) {
 A very simple cli.`,
 		SilenceUsage: true,
+		PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
+			if err := log.InitLogger(); err != nil {
+				return err
+			}
+			return nil
```
```suggestion
			return log.InitLogger()
```

enrichman commented on pkg/cli/cli.go:
```diff
@@ -57,5 +65,8 @@ A very simple cli.`,
 	)
 
+	rootCmd.PersistentFlags().IntVarP(&log.LogLevel, "verbosity", "v", 0, "level of log verbosity")
+	rootCmd.PersistentFlags().StringVarP(&log.LogFileName, "print", "p", "", "print logs to file")
```
Hm, `print` is a bit confusing. I would use another name. What do you think of `logfile` or something like that?

torchiaf replied:
Renamed as `--log-file, -l`

enrichman commented on pkg/log/logger.go:
```diff
@@ -0,0 +1,95 @@
+
+func InitLogger() error {
+
+	ctx = context.Background()
```
I think we should accept a context from the args instead of creating a new one here.

enrichman commented on pkg/log/logger.go:
```diff
@@ -0,0 +1,95 @@
+	levelFatal: "FATAL",
+}
+
+var LogFileName string
+var LogLevel int
```
Hm, globals are not very nice. I would pass those configurations variable in the InitLogger func as well, or in a config struct, for example.

enrichman reviewed (changes requested):
I left some comments, but we can go through them today. :+1:

torchiaf committed:
Update launch step

Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

torchiaf committed:
Update pkg/cli/cli.go

Co-authored-by: Enrico Candino <enrico.candino@gmail.com>

torchiaf committed:
Change logFile flag name

Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

torchiaf committed:
Add logger custom handler to print colors

Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

torchiaf committed:
Remove global log configs

Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

torchiaf committed:
Use command context

Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

torchiaf commented:
Thanks @enrichman , I just pushed the requested changes and added some pics from log file/stdout to the description.
I also added colored output depending on log level. The colors hexa strings are removed before sending output to the log file; It works but need confirmation if it's the correct approach or it could be implemented in a better way.

enrichman commented on pkg/cli/cli.go:
```diff
@@ -7,15 +7,22 @@ import (
 	"k8s.io/client-go/dynamic"
 	"k8s.io/client-go/kubernetes"
+
+	log "github.com/torchiaf/kubectl-rancherx/pkg/log"
```
```suggestion
	"github.com/torchiaf/kubectl-rancherx/pkg/log"
```

enrichman commented on pkg/cli/cli.go:
```diff
@@ -57,5 +64,8 @@ A very simple cli.`,
 	)
 
+	rootCmd.PersistentFlags().IntVarP(&cfg.LogLevel, "verbosity", "v", 0, "level of log verbosity")
+	rootCmd.PersistentFlags().StringVarP(&cfg.LogFileName, "log-file", "l", "", "print logs to file")
```
`l` is a bit "obscure", I would just use the `log-file` flag.

```suggestion
	rootCmd.PersistentFlags().StringVar(&cfg.LogFileName, "log-file", "", "print logs to file")
```

enrichman commented on pkg/log/logger.go:
```diff
@@ -0,0 +1,168 @@
+
+func (e fileWriter) Write(p []byte) (int, error) {
+
+	re := regexp.MustCompile(`\033[[0-9;]*m`)
```
The `MustCompile` could panic. This should be used just outside the func to declare a variable and compile the regex just once:

```go
var myRegex = regexp.MustCompile(`\033[[0-9;]*m`)

func (e fileWriter) Write(p []byte) (int, error) {
    newStr := myRegex.ReplaceAllString(string(p), "")
    // ...
}


```

enrichman commented on pkg/log/logger.go:
```diff
@@ -0,0 +1,168 @@
+type LogConfig struct {
+	LogLevel    int
+	LogFileName string
+}
```
The package is called `log`, and because of this you should avoid to prefix with the same name `Log`: it will "stutter" (i.e.: `log.LogConfig`).

You can just use Config, and this will be `log.Config`.

Also because now the logs are colorized I would add one more config/flag like `ColorsEnabled`/`--no-color`. Some terminals could have problems, and also if we output to a file I guess we should skip (probably it's already done like this).

enrichman commented on pkg/log/logger.go:
```diff
@@ -0,0 +1,168 @@
+	LogFileName string
+}
+
+var ctx context.Context
```
This can be dropped.

```suggestion
```

torchiaf replied:
The `ctx` var is used in log functions below:

```
func Trace(msg string, args ...any) {
	logger.Log(ctx, levelTrace, msg, args...)
}

func Debug(msg string, args ...any) {
	logger.DebugContext(ctx, msg, args...)
}

...
```

Should we keep the global var or add the context as input to the log functions? For instance:

```
func Trace(ctx context, msg string, args ...any) {
	logger.Log(ctx, levelTrace, msg, args...)
}
```

enrichman commented on pkg/log/logger.go:
```diff
@@ -0,0 +1,168 @@
+var ctx context.Context
+var logger *slog.Logger
+
+func InitLogger(context context.Context, cfg *LogConfig) error {
+
+	ctx = context
```
```suggestion
func InitLogger(ctx context.Context, cfg *LogConfig) error {
```

enrichman commented on pkg/log/logger.go:
```diff
@@ -0,0 +1,168 @@
+}
+
+var ctx context.Context
+var logger *slog.Logger
```
I would drop this global. Otherwise it's not possible to create different loggers.

enrichman commented on pkg/log/logger.go:
```diff
@@ -0,0 +1,168 @@
+	return nil
+}
+
+func Trace(msg string, args ...any) {
+	logger.Log(ctx, levelTrace, msg, args...)
+}
```
I would drop these funcs. One we have the `slog.SetDefault` we can just call `slog.Info("Info message")`.

torchiaf replied:
I think we should keep those functions. For custom log `levels` we need to specify the `levelTrace` which is a local variable.

enrichman commented on pkg/manager/rest.go:
```diff
@@ -21,9 +24,25 @@ func List[T runtime.Object](ctx context.Context, client *rest.RESTClient, resour
 		Into(obj)
 
 	if err != nil {
+		log.Trace("manager.List",
+			slog.Group("args",
+				"resource", resource,
+				"namespace", namespace,
+				"err", err,
+			),
+		)
```
Should this be a `log.Error` instead of Trace?

Also I would not Group everything. I like the idea of grouping the `args` and print `args.resource`, `args.namespace`, but I would not print the error in the args.

And the log message should be meaningful, so not `"manager.List"` but something like "error listing resource".

enrichman replied:
It should be also possible to extract the Group, or use a WithXXX

torchiaf replied:
Let's discuss about `Group` / `With` offline.

enrichman reviewed (changes requested):
Some more comments

torchiaf committed:
Update pkg/cli/cli.go - remove import name

Co-authored-by: Enrico Candino <enrico.candino@gmail.com>

torchiaf committed:
Update pkg/cli/cli.go - remove log flag shortcut

Co-authored-by: Enrico Candino <enrico.candino@gmail.com>

torchiaf committed:
Update pkg/log/logger.go - remove unused var

Co-authored-by: Enrico Candino <enrico.candino@gmail.com>

torchiaf committed:
Update pkg/log/logger.go - remove ctx assignement

Co-authored-by: Enrico Candino <enrico.candino@gmail.com>

torchiaf committed:
Add context to logger functions

Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

torchiaf committed:
Declare global colorRegex

Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

torchiaf committed:
Rename LogConfig

Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

torchiaf committed:
Refactoring log calls; Remove default logger as global var

Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

torchiaf committed:
Add --no-color flag

Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

enrichman reviewed (approved):
LGTM, I just wanted to show you about the `log` package from Charm: https://github.com/charmbracelet/log

They do awesome things, and this is one of them. :)
